### PR TITLE
feat: Move bounded_vec into the noir stdlib

### LINKED ISSUE #4020: Move BoundVec into the stdlib
### Problem

The Aztec team has been leaning heavily in a custom data type called [BoundedVec](https://github.com/AztecProtocol/aztec-packages/blob/7002a332da3bb9a75d5164a068a2bd9ea1ad211a/yarn-project/noir-protocol-circuits/src/crates/types/src/utils/bounded_vec.nr#L1) that is an attempt to mock slice/vec functionality using arrays, where essentially you can add to a Vec but that Vec has a max capacity. 

### Happy Case

Move `BoundedVec` into the stdlib for those who would like to use more complex slice/Vec logic. 

Copy the BoundedVec implementation from https://github.com/AztecProtocol/aztec-packages/blob/1c8ea497fb1d64da64cb240917a60d57bd1efef8/yarn-project/noir-protocol-circuits/src/crates/types/src/utils/bounded_vec.nr#L1

### Alternatives Considered

_No response_

### Additional Context

_No response_

### Would you like to submit a PR for this Issue?

No

### Support Needs

_No response_

### PR DESCRIPTION
**Pull request by kevaundray**
# Description

Partially resolves #4020. 

Need to then remove from `aztec-packages` and move to this implementation.

## Problem\*

Resolves 

## Documentation\*

Check one:
- [ ] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[Exceptional Case]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [ ] I have tested the changes locally.
- [ ] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: noir_stdlib/src/bounded_vec.nr
Comment (TomAFrench) [inline noir_stdlib/src/bounded_vec.nr:1]: This should be moved into `collections` to sit with `Vec`

**Review by vezenovm — COMMENTED** [review]
We should just move the tests in stdlib to their own integration tests

**Comment by vezenovm** [comment]
@TomAFrench Moved the tests to `noir_test_success`. I also put `BoundedVec` in the prelude.

File: noir_stdlib/src/bounded_vec.nr [hunk 1]
```diff
@@ -0,0 +1,194 @@
+struct BoundedVec<T, MaxLen> {
+    storage: [T; MaxLen],
+    // TODO: change this to return a u64 as Noir now
+    // uses u64 for indexing
+    len: Field,
```
[the remark this answers is not in the document]
Comment (vezenovm) [reply noir_stdlib/src/bounded_vec.nr]: Oops forgot about this comment. I will update

[the remark this answers is not in the document]
Comment (vezenovm) [reply noir_stdlib/src/bounded_vec.nr]: It does look like arrays also still return a `Field` when calling `.len()`. Perhaps we actually switch this in a separate PR to have all collections follow the same paradigm.

**Review by jfecher — APPROVED** [review]
LGTM. I'm fine with `BoundedVec` in the stdlib as well, but we should wait before merging to see if @TomAFrench's concerns are resolved.

File: noir_stdlib/src/bounded_vec.nr (hunk 1 above)
[the remark this answers is not in the document]
Comment (TomAFrench) [reply noir_stdlib/src/bounded_vec.nr]: I'm fine with holding off on this until we handle all array-ish lengths at the same time.

**Review by TomAFrench — APPROVED** [review]
Nits on naming to be more rust-like. Not sure I would place this in the prelude but I guess if `Vec` is there then it makes sense.

File: noir_stdlib/src/collections/bounded_vec.nr [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        MaxLen
+    }
+
+    // This is a intermediate method, while we don't have an
+    // .extend method
+    pub fn storage(self) -> [T; MaxLen] {
+        self.storage
+    }
+
+    pub fn push_array<Len>(&mut self, array: [T; Len]) {
```
[the remark this answers is not in the document]
Comment (vezenovm) [reply noir_stdlib/src/collections/bounded_vec.nr:44]: I updated to `extend_from_array` instead as `extend_array` was a little confusing to me.

File: noir_stdlib/src/collections/bounded_vec.nr [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        let new_len = self.len + array.len();
+        assert(new_len as u64 <= MaxLen as u64, "push_array out of bounds");
+        for i in 0..array.len() {
+            self.storage[self.len + i] = array[i];
+        }
+        self.len = new_len;
+    }
+
+    pub fn push_vec<Len>(&mut self, vec: BoundedVec<T, Len>) {
```
[the remark this answers is not in the document]
Comment (vezenovm) [reply noir_stdlib/src/collections/bounded_vec.nr:53]: Same here https://github.com/noir-lang/noir/pull/4197#discussion_r1470157494

**Pull request merged**